### PR TITLE
Cknieling/seed UI

### PR DESCRIPTION
**Pull request by chr11115**
Add UI to set seed in assign node

### CODE REVIEW

File: src/palladio/NodeParameter.h [hunk 1]
```diff
@@ -144,6 +145,24 @@ const auto setStartRule = [](OP_Node* node, const std::wstring& s, fpreal t) {
 	node->setString(val, CH_STRING_LITERAL, START_RULE.getToken(), 0, t);
 };
 
+// -- SEED
+static PRM_Name SEED("randomSeed", "Seed");
+static PRM_Range SEED_RANGE(PRM_RANGE_RESTRICTED, INT32_MIN, PRM_RANGE_RESTRICTED, INT32_MAX);
+static PRM_Name OVERRIDE_SEED("overrideRandomSeed", "Override");
+const std::string SEED_HELP = "Sets value for primitive attribute '" + PLD_RANDOM_SEED.toStdString() + "' if override is checked";
```
Comment (mistafunk) [inline src/palladio/NodeParameter.h:152]:
```suggestion
const std::string SEED_HELP = "Sets the random seed for all input primitives (attribute '" + PLD_RANDOM_SEED.toStdString() + "'). This value is ignored if the primitive attribute already exists and "override" is not checked. By default, the random seed is derived from the primitive position.";
```

File: src/palladio/NodeParameter.h [hunk 2]
```diff
@@ -159,7 +178,13 @@ static PRM_Template PARAM_TEMPLATES[] = {
                      START_RULE_HELP.c_str()),
         PRM_Template(PRM_ORD, PRM_Template::PRM_EXPORT_MAX, 1, &CommonNodeParams::LOG_LEVEL,
                      &CommonNodeParams::DEFAULT_LOG_LEVEL, &CommonNodeParams::logLevelMenu),
-        PRM_Template()};
+        PRM_Template(PRM_INT | PRM_TYPE_JOIN_NEXT | PRM_TYPE_PLAIN, 1, &SEED, PRMzeroDefaults, nullptr, &SEED_RANGE,
+                     PRM_Callback(), nullptr,
+                     1, SEED_HELP.c_str()),
+        PRM_Template(PRM_TOGGLE, 1, &OVERRIDE_SEED, PRMzeroDefaults, nullptr, nullptr, PRM_Callback(), nullptr, 1,
+                     SEED_HELP.c_str()),
```
Comment (mistafunk) [inline src/palladio/NodeParameter.h:185]: Use a different help text: "Force setting the random seed even if the corresponding primitive attribute already exists."

**Review by mistafunk — CHANGES_REQUESTED** [review]
LGTM, just some minor improvement suggestions.

File: src/palladio/NodeParameter.cpp [hunk 3]
```diff
@@ -223,6 +224,16 @@ void buildStartRuleMenu(void* data, PRM_Name* theMenu, int theMaxSize, const PRM
 	}
 }
 
+int generateNewSeed(void* data, int, fpreal32 t, const PRM_Template*) {
+	auto* node = static_cast<SOPAssign*>(data);
+	std::random_device rd;
+	std::uniform_int_distribution<int> dist(INT32_MIN, INT32_MAX);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/NodeParameter.cpp:230]: replaced with std::numeric_limits<int>::lowest()/::max()

File: src/palladio/NodeParameter.h (hunk 2 above)
Comment (chr11115) [reply src/palladio/NodeParameter.h]: changed help text

File: src/palladio/NodeParameter.h [hunk 4]
```diff
@@ -144,6 +145,30 @@ const auto setStartRule = [](OP_Node* node, const std::wstring& s, fpreal t) {
...
+const std::string OVERRIDE_SEED_HELP =
+        "Force setting the random seed even if the corresponding primitive attribute already exists.";
+
+const auto getSeed = [](const OP_Node* node, fpreal t) -> int {
+	return node->evalInt(SEED.getToken(), 0, t);
+};
+
+const auto getOverrideSeed = [](const OP_Node* node, fpreal t) -> bool {
+	return static_cast<bool>(node->evalInt(OVERRIDE_SEED.getToken(), 0, t));
+};
```
Comment (mistafunk) [inline src/palladio/NodeParameter.h:166]: There's no actual need for lambdas here. Consider turning them into functions.

**Review by mistafunk — CHANGES_REQUESTED** [review]
LGTM, one minor thing... feel free to directly merge afterwards.

File: src/palladio/NodeParameter.h (hunk 4 above)
Comment (chr11115) [reply src/palladio/NodeParameter.h]: replaced lambdas with real functions and separated body and declaration in cpp/h respectively.
Also sorted function bodys in cpp to match declaration order in h file.

**Pull request merged**